python/semantic_kernel/agents/chat_completion_agent.py

```diff
@@ -0,0 +1,204 @@
+from semantic_kernel.contents.streaming_chat_message_content import StreamingChatMessageContent
+from semantic_kernel.contents.utils.author_role import AuthorRole
+from semantic_kernel.exceptions import KernelServiceNotFoundError
+from semantic_kernel.functions.kernel_arguments import KernelArguments
+from semantic_kernel.utils.experimental_decorator import experimental_class
+
+if TYPE_CHECKING:
+    from semantic_kernel.kernel import Kernel
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+@experimental_class
+class ChatCompletionAgent(ChatHistoryKernelAgent):
+    """A KernelAgent specialization based on ChatCompletionClientBase.
+
+    Note: enable `function_choice_behavior` on the PromptExecutionSettings to enable function
+    choice behavior which allows the kernel to utilize plugins and functions registered in
+    the kernel.
+    """
+
+    service_id: str | None = Field(default=DEFAULT_SERVICE_NAME)
+    execution_settings: PromptExecutionSettings | None = None
+
+    def __init__(
+        self,
+        service_id: str | None = None,
+        name: str | None = None,
+        id: str | None = None,
+        description: str | None = None,
+        instructions: str | None = None,
+        execution_settings: PromptExecutionSettings | None = None,
+    ) -> None:
+        """Initialize a new instance of ChatCompletionAgent.
+
+        Args:
+            service_id: The service id for the chat completion service. (optional) If not provided,
+                the default service name `default` will be used.
+            name: The name of the agent. (optional)
+            id: The unique identifier for the agent. (optional) If not provided,
+                a unique GUID will be generated.
+            description: The description of the agent. (optional)
+            instructions: The instructions for the agent. (optional)
+            execution_settings: The execution settings for the agent. (optional)
+        """
+        super().__init__(name=name, instructions=instructions, id=id, description=description)
+        self.service_id = service_id
+        self.execution_settings = execution_settings
+
+    @override
+    async def invoke(self, kernel: "Kernel", history: ChatHistory) -> AsyncIterable[ChatMessageContent]:  # type: ignore
+        """Invoke the chat history handler.
+
+        Args:
+            kernel: The kernel instance.
+            history: The chat history.
+
+        Returns:
+            An async iterable of ChatMessageContent.
+        """
+        # Get the chat completion service
+        service = kernel.get_service(service_id=self.service_id, type=ChatCompletionClientBase)
+
+        if not service:
+            raise KernelServiceNotFoundError(f"Chat completion service not found with service_id: {self.service_id}")
+
+        # Cast the service to the expected type - satisfy type checking
+        chat_completion_service = cast(ChatCompletionClientBase, service)
```

Comment:
this should be a assert isinstance, casting can have side effects... and the assert can be skipped by going to optimized mode